### ISSUE TITLE
guildMemberSpeaking event no longer working?

### ISSUE THREAD

**Issue by tacodan** [discordjs/discord.js#3524]
**Please describe the problem you are having in as much detail as possible:**

All of a sudden the "guildMemberSpeaking" stopped working. The bot was running did not crash, no errors, all of a sudden the event stop working. It will however emit once when a user starts talking but never will again for that user.

The following is an example of my entire test bot to show the issue:
```js
const path = require('path');
SERVER_DIST = (path.basename(__filename) == "bot.js" ? "prod" : "beta");
CONFIG = require('./config.js').loadVar(SERVER_DIST);

const Discord = require('discord.js');
const client = new Discord.Client();

client.on('error', console.error);

client.on('ready', () => {
	console.log(`Logged in as ${client.user.tag}!`);
	let voiceChan = client.channels.get(CONFIG.ids.mainVoiceChanID);
	voiceChan.join().then(connection => {
		console.log('Connected to voice channel.');
	}).catch(err => console.log(err)); 
});

client.on('guildMemberSpeaking', (member, speaking) => {
	if(speaking) {
		console.log(member.displayName + ' started talking.');    
	}
	else {
		console.log(member.displayName + ' stopped talking');
	}
});

client.login(CONFIG.discordApi.token);
```

**Further details:**
- discord.js version: 11.5.1
- Node.js version: v10.15.3
- Operating system: CentOS release 6.10 (Final)
- Priority this issue should have – please be realistic and elaborate if possible: We use this event to track voice activity in a voice channel, this is very important to this bots usage.

- [ ] I have also tested the issue on latest master, commit hash:

**Comment by dragonbane0** [comment]
Can confirm this started happening for us yesterday as well and really killed us. I assume a discord API change?

**Comment by TreZc0** [comment]
This is becoming a rather severe issue for us, is tere any fix on the horizon?

**Comment by maxymoo22** [comment]
The exact same issue happened for me as well. At first I thought it was a bug in my code, but now I am nearly certain it isn't. Unusual. A fix would be greatly appreciated. :-)

**Comment by njohns67** [comment]
Same issue here. The connection.on("speaking") event seems to fire twice immediately and then never again. I've been debugging for hours trying to figure out what I did wrong before finally finding this thread.

**Comment by Monbrey** [comment]
Is everyone here encountering this issue only on stable? Has anyone had the same issue on the master branch of djs?

**Comment by njohns67** [comment]
@Monbrey I can't try it with the master branch atm but I can tomorrow. If you want to try it yourself I'm sure we'd all appreciate it. If not I'll let you know tomorrow

**Comment by Monbrey** [comment]
> `Syntax Error: Unexpected token * at src/structures/interfaces/Collector.js:203`

This suggests your version of Node is much too old - I believe support for async iterators was added in Node 10 which the master branch requires.

**Comment by Tenpi** [comment]
Same problem on master as well. Seems like there was an api change

**Comment by tacodan** [comment]
I know this feature isn't officially supported as it has been pointed out to me on discord but this feature is really important to my bot. I'd really appreciate it of someone could possibly look into it and determine if maybe there is a work around.  I understand this might not be possible but either way I'd like to say in advance and that your efforts are greatly appreciated.

**Comment by DevYukine** [comment]
As you already pointed out this feature is not "official" supported by discord because recieving voice is completely undocumented and everything undocumented is subject to changes at any point without any notice. This is more like a legacy feature because it used to be (in-official) supported but discord decided against documenting it and with that d.js should not support it anymore. Imo we should actuall deprecate this and remove it in the next major release because it might break at any point without a reliable way to fix that beside testing/reverse engineering against the api (which is forbidden btw :^) ). Im sorry that this might break the "really important" feature but its just legacy  & maintaining this is just a pain at this point.

**Comment by dragonbane0** [comment]
That's unfortunate to hear. Our bot literally relies on being able to capture the isSpeaking event to properly show on our streaming layouts who is currently speaking or not. This is a major convenience feature for the viewers and currently is broken in the sense that everyone is showcased as speaking all the time. So this being supported was a major draw for discord.js.

However I understand if maintaining something that requires constant reverse engineering since discord wastes no effort to document and officially support it is too much to ask for.

If it could be looked into in case it is something simple it would be appreciated, else I try it with a fork I guess. If it gets removed entirely I hope at least getting the members who are inside a voice channel is still gonna work

**Comment by TreZc0** [comment]
To be frank, I'm shocked.
Shocked that the usually so community friendly and committed discord team wants to cut a feature that's actively used in the community and has good reason to exist. We ditched Skype after all cause discord offered so much more convenience and was so much nicer to develop for. Now these developers are kinda getting left hanging.... A mindset I don't particularly understand. 

In all honesty, I get you aiming at us using the discord toolkit voice display, but in it's current, barely responsive and customizable state, it's simple of no use in any kind of proper, somewhat appealing production..

Please, reconsider this..

**Comment by amishshah** [comment]
It seems that no speaking events are received after the first one from the voice gateway. I've asked for some more information on this :+1:

**Comment by tacodan** [comment]
> 
> 
> It seems that no speaking events are received after the first one from the voice gateway. I've asked for some more information on this 👍

Thank you very much for your time in looking into this amishshah. :)

**Comment by dragonbane0** [comment]
Co-signed. Thank you very much amishshah if you can figure something out. As simple as it is, it really allowed all sorts of cool stuff to know when someone speaks. Really strange discord's documentation includes some voice stuff, but not this particular field

**Comment by dragonbane0** [comment]
This is really weird. If you open discord in the browser where the voice connection seemingly uses the webrtc protocol instead of UDP, there are still isSpeaking events coming through for 0 and 1. Feels like a discord bug tbh, but the client can handle it, so it doesn't rely on it to determine whether someone is speaking I guess

**Comment by njohns67** [comment]
As a side issue, has anyone else had issues with `connection.createPCMStream()` after connecting to a voice channel? When `on("speaking")` broke I tried moving my voice receive code outside that block since it runs off the same `connnection` returned from joining the voice channel but it doesn't seem to be working.

**Comment by dragonbane0** [comment]
Capturing voice only works if you send some audio first since a while. The latest master fixed this by making the bot play some silence upon joining a voice channel. Also in the latest stable release, due the onSpeaking event never coming through with a false value, the streams you create with createPCMStream() might not get destroyed properly when the user stops speaking

**Comment by njohns67** [comment]
Thanks I forgot to include my code to play a sound. I'll keep trying to workaround the on.("speaking") bug

**Comment by jhgg** [comment]
We (discord) now only send speaking event once per WS connection - as our clients dont need this event beyond doing initial ssrc => user_id association.

There is a work-around that involves synthesizing the speaking events from UDP packet flow (which is what the native client does.) 

See this example psuedo-ish code for more details:

```js
const speakingMembers = new Map();
const SPEAKING_DELAY = 250;

onRtpPacket(packet) {
    const userId = someSsrcToUserIdMapping.get(packet.ssrc);
    if (userId == null) return;
    const currentSpeakingTimeoutRef = speakingMembers.get(userId);
    if (currentSpeakingTimeoutRef == null) {
        this.emit('speaking', userId);
    } else {
        clearTimeout(currentSpeakingTimeoutRef);
    }
    newSpeakingTimeoutRef = setTimeout(() => { this.emit('stopSpeaking', userId);  speakingMembers.delete(userId); }, SPEAKING_DELAY);
    speakingMembers.set(userId, newSpeakingTimeoutRef); 
}
```

**Comment by Kunikita** [comment]
The VoiceConnection.on listen event is still broken. Just confirmed this. I may have a code solution to share with everyone soon. I need to code and test it real fast first... stand by for that.

**Comment by dragonbane0** [comment]
I couldn't wait for this to get fixed, since we actively use this feature in a relatively big production. So I investigated a bit and indeed inspecting the UDP packages seems to be the way to go. Here is my spin on a quick fix I made 2 days ago, which surprisingly resembles jhgg's version pretty much. 

This is designed for the latest stable version of discord.js, but it can be easily translated I'm sure.

src/client/voice/receiver/VoiceReceiver.js after L30 add:
` this.speakingTimeouts = new Map();`

Before L52 add:
```
if (this.speakingTimeouts.get(ssrc)) {
  clearTimeout(this.speakingTimeouts.get(ssrc));
  this.speakingTimeouts.delete(ssrc);
}
else {
  this.voiceConnection.onSpeaking({user_id: user.id, ssrc: ssrc, speaking: true});
}   
let speakingTimer = setTimeout(() => {
  try {
    this.voiceConnection.onSpeaking({ user_id: user.id, ssrc: ssrc, speaking: false });
    this.speakingTimeouts.delete(ssrc);
  }
  catch (ex) {
    console.log("Connection already closed");
  }
}, 50);

this.speakingTimeouts.set(ssrc, speakingTimer);
```
This gives you the connection.on('speaking') and client.on('guildMemberSpeaking') events back in a pretty much perfectly working state, as well as fix the streams not ending when someone stops speaking if you record them. 

Remember you need to play some silence first in order for the bot to receive voice packages properly.

**Comment by Kunikita** [comment]
Just got it working again. Here's how I'm getting it to work...

`connection.on('speaking', (user) => {
            if(user.bot === true) return;

            // Creating user's voice stream...
            try {
                // HACK: Trick to get stream open "Discord.js bug", or else listening doesn't work!!!!
                class Silence extends Readable {
                    _read() { this.push(Buffer.from([0xF8, 0xFF, 0xFE])); }
                } connection.playOpusStream(new Silence());

                const delayTime = 250;
                let startTime = new Date().getTime();
                const receiver = connection.createReceiver();
                const voiceStream = receiver.createPCMStream(user);

                voiceStream.on('data', (chunk) => {
                    if(speakingPaused()) {
                        startSpeaking();
                        resumeSpeaking();
                    }
                    startTime = new Date().getTime();
                });

                // eslint-disable-next-line no-inner-declarations
                function speakingPaused() {
                    const time = new Date().getTime();
                    const deltaTime = time - startTime;
                    return deltaTime > delayTime;
                }

                // eslint-disable-next-line no-inner-declarations
                function resumeSpeaking() {
                    setTimeout(() => {
                        if(speakingPaused()) {
                            stopSpeaking(); return;
                        } resumeSpeaking();
                    }, delayTime);
                }

                // eslint-disable-next-line no-inner-declarations
                function startSpeaking() {
                    console.log(`${user.username} has started speaking...`);
                }

                // eslint-disable-next-line no-inner-declarations
                function stopSpeaking() {
                    console.log(`${user.username} has stopped speaking...`);
                }
            }
            catch(err) { console.error(err); return; }
        });
    }).catch(err => {
        console.error(err);
    });`

**Comment by E0HH** [comment]
The guildMemberSpeaking event is somehow working again, even though jhgg confirmed its only supposed to be firing only once. This makes it hard because it seems that I am making the event fire twice now if I implement any changes, or breaks it again all together, because I assume it is cancelling itself out. If this is only temporary, and is supposed to break again, what is the most ideal way to handle double events on master v12 until we find out? I am trying to handle the events inside the packetHandler push function, since it seems the most logical place to get the user and ssrc. Sorry to piggyback off the post, but this seems like the best place to secure an answer for myself and others trying to fix their bots.

**Comment by njohns67** [comment]
@dragonbane0 This does seem to properly emit the on("speaking") event but I'm running into some bugs that I didn't have previously. It seems the event fires twice in a row, causing a pcmStream error if you are trying to use the incoming voice data.

```
newMember.voiceChannel.join().catch(console.error).then(async function(connection) {
        connection.on("disconnect", () => console.log("Disconnecting dispatcher"))
        console.log("Connected")
        if(first){
           class Silence extends Readable {
               _read(){
                   this.push(Buffer.from([0xF8, 0xFF, 0xFE]))
               }
            }
           connection.playOpusStream(new Silence())
            first = false
        }
        const receiver = connection.createReceiver()
        connection.on('speaking', (user, speaking) => {
            console.log("Speaking")
            if (!speaking){
                return
            }
            const audioStream = receiver.createPCMStream(user)
           //More code to do stuff with the audioStream
        })
    })
})
```
This is my code minus any extraneous bits for readability's sake. Currently my console log looks like: ```
$ node bot.js
$ speaking
$ speaking
$  if (this.pcmStreams.get(user.id)) throw new Error('There is already an existing stream for that user.');
                                      ^

Error: There is already an existing stream for that user.
```
As you can see, it emits the event twice before crashing. Through testing I've found that it only emits twice the first time you speak. I'm guessing the first time is the event properly being emitted (as it was before everything broke) and the second time is your code emitting the event with subsequent emits being your code as well. 

I was able to somewhat workaround this by modifing VoiceReceiver.js L161 to return the existing pcmStream, but this seems like a bandaid. It also cause the voice data to be corrupted.

**Comment by dragonbane0** [comment]
@njohns67 Yeah your assumption is correct. The first time would trigger twice. I guess the code can be altered to suppress the actual speaking event entirely now besides doing the user--->ssrc mapping

**Comment by Kunikita** [comment]
njohns67, dragonbane0 - Here's how I fixed it from triggering twice...

Adding the following after L32...
const speakingUsers = [];

Than make changes to the first else condition. See my changes bellow...

```
if (this.speakingTimeouts.get(ssrc)) {
          clearTimeout(this.speakingTimeouts.get(ssrc));
          this.speakingTimeouts.delete(ssrc);
}
else if(speakingUsers.includes(user.id)) {
          this.voiceConnection.onSpeaking({user_id: user.id, ssrc: ssrc, speaking: true});
} else {
          speakingUsers.push(user.id);
}
```

**Comment by Kunikita** [comment]
@njohns67 I got it to stop playing twice by adding a new ws.on event to VoiceConnection.js, around LN 368; which just sets this.ssrcMap. Then, instead of calling this.emit('speaking'... from VoiceWebSocket.js, I'm calling that new event.

Hope that helps and doesn't cause any problems for you. It's working for me!

**Comment by njohns67** [comment]
Thank you for all your help. I ended up sticking with the modified VoiceReceiver.js code and simply returning the stream if there's an error. This may causes issues I'm unaware of but none have come up and until this is officially fixed I'm sticking with my bandaid for time's sake.

**Comment by tacodan** [comment]
I still haven't been able to get this working, I tried to apply the changes that dragonbane0 provided but still nothing :(

Can anyone help me out?

**Comment by Kunikita** [comment]
@tacodan I got it to work. Here's what I did. Hope it helps!

VoiceReceiver.js - Add nJohn67's code...

VoiceWebSocket.js - change the emit event from 'speaking' to 'startSpeaking'...

```
...
case Constants.VoiceOPCodes.SPEAKING:
        ...
        this.emit('startSpeaking', packet.d);
        break;
...
```
VoiceConnection.js - Add the 'startSpeaking' event and the ws.on and ws.remove listeners...
```
...
onStartSpeaking({ user_id, ssrc, speaking }) {
    const user = this.client.users.get(user_id);
    this.ssrcMap.set(+ssrc, user);
}
...
      ws.removeAllListeners('speaking');
      ws.removeAllListeners('startSpeaking');
    }
...
    ws.on('speaking', this.onSpeaking.bind(this));
    ws.on('startSpeaking', this.onStartSpeaking.bind(this));
  }
...
```

**Comment by tacodan** [comment]
> 
> 
> @tacodan I got it to work. Here's what I did. Hope it helps! This also keeps 'listen' from firing twice in the beginning...
> 
> VoiceReceiver.js - Add nJohn67's code...
> 
> VoiceWebSocket.js - change the emit event from 'speaking' to 'startSpeaking'...
> 
> ```
> ...
> case Constants.VoiceOPCodes.SPEAKING:
>         ...
>         this.emit('startSpeaking', packet.d);
>         break;
> ...
> ```
> 
> VoiceConnection.js - Add the 'startSpeaking' event and the ws.on and ws.remove listeners...
> 
> ```
> ...
> onStartSpeaking({ user_id, ssrc, speaking }) {
>     const user = this.client.users.get(user_id);
>     this.ssrcMap.set(+ssrc, user);
> }
> ...
>       ws.removeAllListeners('speaking');
>       ws.removeAllListeners('startSpeaking');
>     }
> ...
>     ws.on('speaking', this.onSpeaking.bind(this));
>     ws.on('startSpeaking', this.onStartSpeaking.bind(this));
>   }
> ...
> ```

do you mean dragonbane0's code for VoiceReceiver.js or?

**Comment by Kunikita** [comment]
@tacodan yes, I mean dragonbane0's code... sorry... my account is bodenbao.

**Comment by bsguedes** [comment]
Are these changes going to be integrated in an official release in discord.js, so the guildMemberSpeaking event returns to its normal behavior? Thanks!

**Comment by CLEMARCx** [comment]
Is this fix still working? Cant get it to work.

**Comment by sillyfrog** [comment]
@tacodan Do you have a fork with theses changes applied? Sounds like it maybe worth making a PR to get this into master as it would be great to have. I can also make a PR, but sounds like you have already done the work.
Cheers.

**Comment by sillyfrog** [comment]
@Kunikita Do you have a fork with theses changes applied? Sounds like it maybe worth making a PR to get this into master as it would be great to have. I can also make a PR, but sounds like you have already done the work.
Cheers.

**Comment by christopherfowers** [comment]
Hate to be "that guy" but has there been any progress on this? Or is this even going to be addressed since it's not technically supported by discord?

**Comment by Nameless9** [comment]
@sillyfrog hello, I was trying to use your fork version. I have some mixed result. Sometimes, the speaking event works, which is great. Other times, none of the events work, let alone, speaking event. Even disconnect event doesn't get triggered when I disconnect the voice connection.

**Comment by christopherfowers** [comment]
Yeah. Haven't had a chance to debug yet but noticed no events working. Was hoping to take a look some time today. I'll try tomorrow.

**Comment by Nameless9** [comment]
can I know if anyone else is experiencing none of the event-handlers on connection not working?

**Comment by christopherfowers** [comment]
@Nameless9 it looks like the bot may have something else going on. Noted that when the bot joins my voice channel, it is broadcasting 
![image](https://user-images.githubusercontent.com/5230453/70192349-81caa380-16c9-11ea-874e-5cdf361d9536.png). Yet there is no point that I have provided any statements for this action in code. 

I assume you may find that you're experiencing something similar. 

`// jshint esversion: 8
const Discord = require("discord.js");
const config = require("./config.json");
const discord = new Discord.Client();

discord.login(config.discordApiToken);

discord.on("ready", () => {
  console.log(`Logged in as ${discord.user.tag}!`);
});

discord.on("message", async msg => {
  if (msg.content.startsWith("#join")) {
    let [command, ...channelName] = msg.content.split(" ");

    if (!msg.guild) {
      return msg.reply(
        "no private service is available in your area at the moment. Please contact a service representative for more details."
      );
    }

    const voiceChannel = msg.guild.channels.find(
      channel => channel.name === channelName.join(" ")
    );

    if (!voiceChannel || voiceChannel.type !== "voice") {
      return msg.reply(
        `I couldn't find the channel ${channelName}. Can you spell?`
      );
    }

    const connection = await voiceChannel.join();
    console.log("Ready!");

    connection.on("speaking", (user, speaking) => {
      console.log(speaking);
      if (speaking) {
        console.log(`${user.username} is speaking`);
      } else {
        console.log(`${user.username} stopped speaking`);
      }
    });
  }
});
`

**Comment by sillyfrog** [comment]
What exact version of `node` are you using? I have hacked something together so you can see it in action in _Docker_. This is based on what I used for all my dev and testing: https://github.com/sillyfrog/discord.js-listenexample

That should be a basis for you to get started, you can then work from there to hopefully debug it. This also shows all of the packages I'm using in both Linux and node (which if missing could be causing issues).

**Comment by ptree44** [comment]
@christopherfowers I'm working with @Nameless9 .. we have that very exact issue. We have tried node-opus and opusscript. We have tried joining channels, leaving, rejoining, creating new broadcasts, etc to no avail.

We tried it on node 10.2 and node 10.16

**Comment by christopherfowers** [comment]
> What exact version of `node` are you using? I have hacked something together so you can see it in action in _Docker_. This is based on what I used for all my dev and testing: https://github.com/sillyfrog/discord.js-listenexample
> 
> That should be a basis for you to get started, you can then work from there to hopefully debug it. This also shows all of the packages I'm using in both Linux and node (which if missing could be causing issues).

I am on windows 10 with node 12.13.0

I do tend to run linux docker containers and I suppose I could attempt this from my mac even. 

I can try to rule out node versions first. Thankfully we have NVM to the rescue. I'll let you know what I find out.

**Comment by christopherfowers** [comment]
Can confirm the Demo you made works @sillyfrog. @ptree44  if you would like some help getting that working let me know. We can sidebar this conversation. Meanwhile going to press on with seeing if node versioning is my issue.

**Comment by christopherfowers** [comment]
@ptree44 can you tell me more about your environment? What OS are you on running on? What packages do you have installed (specifically node packages)?

**Comment by ptree44** [comment]
> @ptree44 can you tell me more about your environment? What OS are you on running on? What packages do you have installed (specifically node packages)?

"dependencies": {
    "@discordjs/uws": "^11.149.1",
    "big-integer": "^1.6.48",
    "body-parser": "1.19.0",
    "brain.js": "^1.2.2",
    "bufferutil": "^4.0.1",
    "cloudscraper": "^4.1.2",
    "discord.js.master": "github:discordjs/discord.js#master",
    "discord.js.stable": "github:discordjs/discord.js#stable",
    "eris": "0.10.0",
    "erlpack": "github:discordapp/erlpack",
    "express": "^4.16.4",
    "firebase": "6.3.4",
    "get-pixels": "^3.3.2",
    "googleapis": "^39.2.0",
    "is-json": "^2.0.1",
    "libsodium-wrappers": "^0.7.6",
    "obs-websocket-js": "^3.1.0",
    "opusscript": "0.0.7",
    "pnpm": "2.25.5",
    "promise": "^8.0.3",
    "request": "2.88.0",
    "timer-node": "1.0.4",
    "tmi.js": "^1.2.1",
    "when": "3.7.8",
    "ytdl-core": "^1.0.3",
    "zlib-sync": "^0.1.6"
  },
  "engines": {
    "node": "10.16.x"
  },


from our package.json

**Comment by Lovinity** [comment]
Quick question relating to this. In the distant past, I recall the speaking event could fire for changes in speaking state even if the bot was not in the voice channel. Analyzing the suggested code changes above, it appears that is no longer the case and the bot absolutely has to be in a voice channel in order to receive speaking states. Is this true? And if so, is there a possible workaround to this, or is it bottlenecked by Discord's API?

**Comment by ptree44** [comment]
@Lovinity The guildMemberSpeaking event supposedly should work for bots outside of the voice channel. I can't get the guildMemberSpeaking event nor speaking event for VoiceConnection to work though, but I do seem to be in the minority.

**Comment by Lovinity** [comment]
@ptree44 They do not work for me either. Earlier in the thread, people mentioned it firing once or twice and that's it. For me, it doesn't ever fire. I debugged by checking the raw event, and raw is not coming up with any speaking nor guildMemberSpeaking events either.

**Comment by Fyko** [comment]
This has since been fixed with #3578.

**Comment by ptree44** [comment]
@Lovinity I tried on multiple node versions and it didn't work for me. I've also tried fresh projects , still didn't work. It's a needed feature in my bot , very disappointed.

**Comment by ptree44** [comment]
@Fyko that did not fix it in my circumstance.

**Comment by Lovinity** [comment]
I too have the latest master commit and it does not work.

**Comment by sillyfrog** [comment]
Discord have changed the way this works. It's not part of the official API, hence the original breaking change. It now works differently in that the bot needs to be in the channel as there are no longer speaking/stop speaking events fired via the web socket (there is just one initial speaking event to tie the various ID's together). I have put together a working example that runs in Docker so you can see it in action, then you can work from there to see what maybe different in your setup: https://github.com/sillyfrog/discord.js-listenexample

**Comment by Kunikita** [comment]
Is everyone okay if I push a change to fix this bug?

**Comment by Monbrey** [comment]
Anyone is welcome to make a PR to the master or 11.5-dev branches, which will be reviewed and merged if valid.

**Comment by SpaceEEC** [comment]
This has been fixed in both relevant branches now:
- master -> https://github.com/discordjs/discord.js/pull/3578 (linked above somewhere)
- 11.5-dev -> https://github.com/discordjs/discord.js/pull/3749

**Comment by jackred** [comment]
Using 11.5-dev ([c955fd00c7ef5835e022db45ac16d8fe24689455](https://github.com/discordjs/discord.js/commit/c955fd00c7ef5835e022db45ac16d8fe24689455))

The event `guildMemberSpeaking` on client and `speaking` are working, but intermittently. 

When I make the bot join the vocal channel, sometimes the events work perfectly, sending true/false every time I start/stop speaking, but most of the time it just do nothing.

I tried on different server, using different bot, and trying alone or with someone. I couldn't find a denominator between the working / not working part.
I thought it was a problem with how I closed the bot, as I originally found some sort of connection with the event working when using a new bot or a new channel, but after more test it doesn't seem like it.

using:
```js
function handleVocalMessage(userOrMember, speaking){
  console.log(userOrMember.displayName || userOrMember.username, "is talking?", speaking);
}

function joinChannel(channel) {
  channel.join().then(d => {
    console.log(`INFO! joined ${d.channel.name}`);
    d.on('speaking', handleVocalMessage);
  });
}

client.on('guildMemberSpeaking', handleVocalMessage);
```

[gist of complete testing code](https://gist.github.com/jackred/cd22e6802226d9bff66d33b6f28c00c6)

![image](https://user-images.githubusercontent.com/32414502/73840574-60dd1680-4810-11ea-9b91-c12509932d1c.png)


![image](https://user-images.githubusercontent.com/32414502/73840549-53c02780-4810-11ea-8ad4-0cadb14ec4ab.png)

**Comment by KilMer56** [comment]
Hey, for me it's still not working on the master branch. In fact, with the simple following code :

`connection.on('speaking', (user, speaking) => {
        console.log(speaking);
});`

The console is only displaying 'true' once and then never work again.

**Comment by theLMGN** [comment]
Seems to be busted again. `Connection.on('speaking',(user,speaking) => {console.log("speaking")})` don't output anythign and debug logging only outputs the one speaking event when the user first speaks after joining the VC.

**Comment by sillyfrog** [comment]
@theLMGN , would you please try running the https://github.com/sillyfrog/discord.js-listenexample example? I have just rebuilt and tested again (so it's against current master), and things are working OK. I would suggest it's most likely dependencies or similar causing issues.

**Comment by SunburntRock89** [comment]
@sillyfrog The script hasn't been updated and doesn't follow the new ChannelManager class change and so doesn't work with current master. It hasn't been updated since December, maybe you forgot to commit?

**Comment by SunburntRock89** [comment]
Followup - changing it to grab the channel from master, it actually does work. There seems to be some kind of link between playing a sound and having the speaking event fire. Strange.

**Comment by dragonbane0** [comment]
Yes you need to "speak" before Discord gives you audio. This has been a change on their side and can not be avoided

**Comment by sillyfrog** [comment]
@SunburntRock89 Thanks, it did indeed need some updates (including pulling from the correct repo)! That's now done should anyone else need it. Cheers.

**Comment by DominusVilicus** [comment]
Can this get fixed?

**Comment by CoderSudaWuda** [comment]
Help we're doomed.

**Comment by CoderSudaWuda** [comment]
No srsly we're doomed.

**Comment by DominusVilicus** [comment]
This still doesn't work in V12? Buggy as hell

**Comment by SunburntRock89** [comment]
What's your issue @DominusVilicus

**Comment by amishshah** [comment]
Is anyone on this thread still experiencing this issue on the `master` branch of discord.js?

**Comment by MaxmaxmaximusGitHub** [comment]
@amishshah 
![image](https://user-images.githubusercontent.com/60403187/83640611-0765d800-a5b5-11ea-9d81-4d164d41c079.png)


![image](https://user-images.githubusercontent.com/60403187/83640626-0f257c80-a5b5-11ea-826e-cc5bce6a672b.png)

**Comment by sillyfrog** [comment]
@MaxmaxmaximusGitHub how are you running this? What version of Node etc? That method should return a timer object (not a number).

**Comment by MaxmaxmaximusGitHub** [comment]
@sillyfrog 
![image](https://user-images.githubusercontent.com/60403187/83643233-73960b00-a5b8-11ea-9514-80c8869d470e.png)


master, and 12 branches,
electron with node 12

**Comment by sillyfrog** [comment]
@MaxmaxmaximusGitHub I'm guessing that electron is messing with the `setTimeout` implementation to make it more browser like (I know basically nothing about electron, so may well be off the mark). The node docs show it should return a `<timeout>` object: https://nodejs.org/api/timers.html#timers_settimeout_callback_delay_args
Running my [example](https://github.com/sillyfrog/discord.js-listenexample) just now (in a fresh Node 12 docker image), adding in a log line:
```
@@ -103,6 +103,7 @@
       }, DISCORD_SPEAKING_DELAY);
       this.speakingTimeouts.set(ssrc, speakingTimeout);
     } else {
+      console.log("speakingTimeout:", speakingTimeout);
       speakingTimeout.refresh();
     }
```
I get this output (a _lot_ as it's on every single packet):
```
speakingTimeout: Timeout {
  _idleTimeout: 250,
  _idlePrev: [TimersList],
  _idleNext: [TimersList],
  _idleStart: 7728,
  _onTimeout: [Function],
  _timerArgs: undefined,
  _repeat: null,
  _destroyed: false,
  [Symbol(refed)]: true,
  [Symbol(asyncId)]: 858,
  [Symbol(triggerId)]: 95
}
```
Hopefully that helps at least point you in the right direction. Cheers.

**Comment by d4sein** [comment]
I'm having a similar issue, the event actually fires every time I talk and stop talking, but I can't receive any stream of audio from it.

```js
    case 'join':
      if (ctx.member.voice.channel) {

        const connection = await ctx.member.voice.channel.join()
        connection.play(new Silence(), { type: 'opus' })

        connection.on('speaking', (user, speaking) => {
          let audioStream = connection.receiver.createStream(user, { mode: 'pcm' })
  
          // Transforms the audio stream into something Dialogflow understands
          ffmpeg(audioStream)
            .inputFormat('s32le')
            .audioFrequency(44100)
            .audioChannels(1)
            .audioCodec('pcm_s16le')
            .format('s16le')
            .on('error', console.error)
            .pipe(fs.createWriteStream(`${user.id}.wav`))
          
          if (!speaking) return
        })
      }
      break
```

It works If I do the same thing outside of `connection.on('speaking', ...)`.
`client.on('guildMemberSpeaking')` also doesn't work.

**Comment by MegaNoam** [comment]
Is there anything new with the issue? @amishshah

**Comment by amishshah** [comment]
Hi there,

We're working on a new implementation of Discord's Voice API that has better playback quality and is more reliable than what we currently support in Discord.js v12 - check it out at https://github.com/discordjs/voice!

The new library solves many of the issues that users are facing, and as part of this, we're dropping built-in support for voice in our next major release. We have a PR (https://github.com/discordjs/discord.js/pull/5402) that adds native support for our new voice library - once this PR is merged, this issue will be closed.

You can still use our new voice library before that PR lands - just take a look at [our music bot example](https://github.com/discordjs/voice/tree/main/examples) to see how to get started upgrading your voice code. By using the boilerplate music player in the example, you can make it even easier to upgrade your code.

Note that the PR above only reduces some of the boilerplate code you'd otherwise have to write - you do not have to wait for the PR to be merged to start using the new voice library.

-----

If you have any questions about this, feel free to:

- [Make an issue](https://github.com/discordjs/voice/issues) if you have found a bug in the new voice library
- Use [GitHub Discussions](https://github.com/discordjs/voice/discussions) or join our [Discord server](https://discord.gg/djs) (we have a new channel, `#djs-new-voice`, specifically for this!) to ask general questions about the library, give feedback on the library, and get support with upgrading to it

-----

**Specific to this issue:**

This event relies on the voice receive feature, which is undocumented by Discord and prone to breaking quite a lot. As it's an undocumented feature, we cannot guarantee support for it.
